crypto/asn1/tasn_dec.c

```diff
@@ -316,7 +316,7 @@ static int asn1_item_embed_d2i(ASN1_VALUE **pval, const unsigned char **in,
             if (tt->flags & ASN1_TFLG_ADB_MASK) {
                 const ASN1_TEMPLATE *seqtt;
                 ASN1_VALUE **pseqval;
-                seqtt = asn1_do_adb(pval, tt, 0);
+                seqtt = asn1_do_adb((const ASN1_VALUE **)pval, tt, 0);
```

Comment:
Are these and other casts of similar type where const is added actually valid?

See: http://c-faq.com/ansi/constmismatch.html

Reply:
Thanks for asking. I've just checked again in `crypto/asn1/tasn_utl.c`:
```
 const ASN1_TEMPLATE *asn1_do_adb(const ASN1_VALUE **pval,
                                 const ASN1_TEMPLATE *tt,
                                 int nullerr)
```
really does not alter (directly or indirectly) `**pval`, i.e., the contents pointed to by `*pval`, so the three casts to `const` I added to `crypto/asn1/tasn_dec.c` are justified. 

Should I add comments making this explicit?

Reply:
My main worry would be that this code changes in the future and the added casts hide something that would be an error. I have no opinion on what the correct thing to do is.

Reply:
This worry is applicable actually also to many many other casts in the OpenSSL library code, also to those that adapt the type interpretation (rather than the constantness) of arguments and results of function calls.

What might be best in these cases is to make the (currently hopefully fulfilled) assumptions by the function users explicit to the function definition, for instance by adding a comment before the line
```
const ASN1_TEMPLATE *asn1_do_adb(const ASN1_VALUE **pval,
```
like the following:
```
/* beware: callers of this function rely in the fact that **pval is not modified */ 
```

Reply:
Thinking a little more about this issue I meanwhile believe that the compiler's complaint at points like the above where a non-const `T**` is passed to a `const T**` parameter is actually wrong and the explanation given in http://c-faq.com/ansi/constmismatch.html is misleading. 

The explanation there pretends to talk about such a function argument passing but in fact gives an example that contains aliasing via two local pointer variables referring to the same pointer. There the situation is different because the actual value pointed to can be accessed in two different ways (namely, a constant and a modifiable way). In contrast, in the case of passing a `T**` pointer to a function that expects (or better to say: promises) `const T**` the compiler will prevent the function body from modifying `T`. The caller can rely in this as long as the type of the function argument is not changed to `T**`. And after the function call has terminated only the non-const `T**` remains, and the calling code is free to modify the value pointed to. `

In our case here it is `asn1_item_embed_d2i()` which has no restrictions modifying *pval`. This function actually does not need the promise given by `asn1_do_adb()`. So the compiler error given without the cast(s) to `const` that I had to insert makes no sense to me.

Reply:
Digging deeper,it turns out that the issue is even more subtle, both the explanation and example given in http://c-faq.com/ansi/constmismatch.html is very misleading, and the compiler was actually right to give a warning here. 

Have a look at the simple and very instructive example at https://stackoverflow.com/questions/45011978/const-causing-incompatible-pointer-type-why-only-for-double-pointers#45013362
to understand that the actual concern is that `asn1_do_adb()` might re-assign `*pval` to a pointer to something that is not modifiable although `asn1_item_embed_d2i()` is entitled to modify it.

Reply:
Interestingly, although it is possible to constify the `pval` parameter of `asn1_do_adb()` also at the first pointer level: `const ASN1_VALUE * const *pval` this unfortunately does not help to avoid the cast(s) to `const` when using this function because the type rules used by the compiler at not smart enough.

Reply:
Yet I've just realized that the issue being discussed here is simply due to inadequate interface design for `asn1_do_adb()`: the `const ASN1_VALUE **pval` can be simplified to  `const ASN1_VALUE *val` and then all the casts to const in question are not needed anymore. I've pushed an update implementing this.

Reply:
Here be dragons!

Isn't `asn1_do_adb()` pretty intricate re embedded types (typically for the INT family of ASN.1 types, defined by `crypto/asn1/x_int64.c`)

Reply:
On the other hand, all tests seem to fly, including the one test that checks this, so maybe I'm over reacting...

Reply:
For this PR I had to work through lots of pretty obscure and mostly undocumented internal ASN.1 stuff, and I'm far from getting the while picture.
However, looking at its code I'm sure that `asn1_do_adb()` handles its arguments and the memory areas they point in purely read-only fashion. In particular, regardless what `adb->adb_cb()` does, the callback can only change the value of the local variable `selector` (but not `*val` nor `*tt`).
And after getting rid of the needless pointer indirection in the first parameter of `asn1_do_adb()`, the compiler confirms this.